### ISSUE TITLE
fs: unable to fsync an open directory

### ISSUE THREAD

**Issue by perillo** [ziglang/zig#15563]
### Zig Version

0.11.0-dev.2935+ec6ffaa1e

### Steps to Reproduce and Observed Behavior

```zig
const std = @import("std");

const debug = std.debug;
const fs = std.fs;
const os = std.os;

const truth = false;

pub fn main() !void {
    var cwd = fs.cwd();

    var dir = if (truth)
        cwd
    else
        try cwd.openDir(".", .{});

    debug.print("fd: {}\n", .{dir.fd});
    try os.fsync(dir.fd);
}
```

```sh
zig run example.zig
```

The output when `truth = false` is:
```
fd: 3
thread 559313 panic: reached unreachable code
/home/manlio/.local/share/sdk/zig/master/lib/std/os.zig:6891:33: 0x20ba8a in fsync (example)
        .BADF, .INVAL, .ROFS => unreachable,

...
```

The output when `truth = true` is:
```
fd: -100
thread 559982 panic: reached unreachable code
/home/manlio/.local/share/sdk/zig/master/lib/std/os.zig:6891:33: 0x20b73a in fsync (example)
        .BADF, .INVAL, .ROFS => unreachable,

...
```

### Expected Behavior

  1. `os.fsync` should not call the syscall, when `fd` is the special value `os.AT.FDCWD` on UNIX systems.
      Not not sure about Windows and WASI.
  3. `fs.openDir` should not open the directory in readonly mode.
      The open mode should be specified in `OpenDirOptions`.
      What is the rationale?

**Comment by matu3ba** [comment]
You should get `EINVAL The fildes argument does not refer to a file on which this operation is possible.` or `ROFS`.
`cwd` is a special file, since it is listed separately and not "merely a dir" https://stackoverflow.com/a/52457592.
Windows does not have this behavior.

1. This is only possible in all cases, if Zig would do a full Linux abstraction over all syscalls to track separately in an enum the state of a file handle for all syscalls or it would be a halfbaked solution. This would additionally require from the user extra-annotations to opt out of this checks for C interoperation. Also take note, that the api surface of the Linux Kernel is huge.
The alternative is force every user to do an additional system call, even though the user does not need the file descriptor to `cwd`.
2. The rationale is that all file access, which is not based on handles is racy. With file handles and you need the parent dir for this, it is not. Thus it makes sense to nudge the developer into getting the parent dir handle to delete the current directory to prevent race conditions from the start.

**Comment by squeek502** [comment]
Related: https://github.com/ziglang/zig/issues/9688

**Comment by perillo** [comment]
@matu3ba Thanks for the comment.

I know the advantages of `openat` and friend.

The problem is that the current API prevents the use of the classic atomic replacement pattern, i.e.,

  - write temporary file
  - fsync temporary file
  - rename temporary file over real file
  - fsync parent directory

Quoted from: https://patchwork.kernel.org/project/linux-fsdevel/patch/c823982d5b46ea888dc1fdf26c067a7aa0f3585f.1490103963.git.osandov@fb.com/.

`std.fs` has atomic file support, but I'm not sure it is equivalent to the classing pattern, since it doesn't use fsync.

It is possible that the implementation wants to rely on the `AT_REPLACE` patch in the linux kernel instead of using fsync?

**Comment by matu3ba** [comment]
> The problem is that the current API prevents the use of the classic atomic replacement pattern, i.e.,
>
>    write temporary file
>    fsync temporary file
>    rename temporary file over real file
>    fsync parent directory

You can still get the directory handle to the `cwd` path with `openDirAbsolute(., ..)` or `openDir(std.fs.cwd, ...)` and likewise do the actions from there.

> It is possible that the implementation wants to rely on the AT_LINK_REPLACE patch in the linux kernel instead of using fsync?

You could use the `linkat` syscall in Linux directly instead of `rename` and `fsync`. There is already a binding to `linkat`.

**Comment by perillo** [comment]
> > The problem is that the current API prevents the use of the classic atomic replacement pattern, i.e.,
> [...]
 
> You can still get the directory handle to the `cwd` path with `openDirAbsolute(..., etc)` or `openDir(std.fs.cwd(), etc)` 

No, you can't.  `OpenDir` always set the `os.O.RDONLY` flag, so `fsync` will fail.

> > It is possible that the implementation wants to rely on the AT_LINK_REPLACE patch in the linux kernel instead of using fsync?
> 
> You could use the `linkat` syscall in Linux directly instead of `rename` and `fsync`. There is already a binding to `linkat`.

Maybe `linkat` ensures atomicity, but does also ensure durability?

**Comment by matu3ba** [comment]
> No, you can't. OpenDir always set the os.O.RDONLY flag, so fsync will fail.

You are correct. Not sure, what the best way to solve this would be.

> Maybe linkat ensures atomicity, but does also ensure durability?

I dont see anything in `man 2 linkat` which indicates that it is temporary.

**Comment by lsculv** [comment]
This is resolved at least on Linux as of Zig 0.13, you can call fsync on a directory like this:
```zig
const std = @import("std");

test {
    const dir = try std.fs.cwd().openDir(".", .{ .iterate = true });
    try std.posix.fsync(dir.fd);
}
```

The `iterate` DirOption is required as it disables the `O.PATH` open option being used. The following code in the `std.fs.openDirZ` is relevant:
https://github.com/ziglang/zig/blob/d72f3d353f771daced78af70c049e3c5075b3529/lib/std/fs/Dir.zig#L1535-L1536

`O_PATH` is _kind of_ documented to have this behavior and it makes sense, but I think for the standard library this behavior is more than a little non-obvious. Why should I need to pass the `iterate` flag to `fsync` my directory? Is this a guarantee? Should it be?

https://www.spinics.net/lists/linux-fsdevel/msg124527.html

> As an alternative, this patch only has syncfs report errors recorded
> in s_wb_err when the file has been opened with O_PATH. Any file opened
> with O_PATH will not have its fsync field defined in its
> file_operations so we can be sure that nothing else will be using its
> f_wb_err
>  
> Note that calling syncfs on an O_PATH descriptor today will return
> -EBADF, so this scheme gives userland a way to tell whether this
> mechanism will work at runtime.

https://lwn.net/Articles/752209/

> I proposed a draft patch earlier this week (which I meant to send as an RFC) that does the latter. It's based on Willy's suggestion to only report errors from the errseq_t when the fd is an O_PATH open. You can't call fsync on an O_PATH open, so that should be safe (though it is horribly non-obvious from a userland API standpoint)

I also don't think getting an `EBADF` from a `fsync` call should give you a guaranteed panic because of getting to unreachable code (in `std.posix.fsync`):
https://github.com/ziglang/zig/blob/d72f3d353f771daced78af70c049e3c5075b3529/lib/std/posix.zig#L7021-L7029
